### PR TITLE
Add regex capture-group parameter to auto convert to non-capture groups

### LINKED ISSUE #11663: [FEA] avoid regexp capture groups automatically in some cases
**Is your feature request related to a problem? Please describe.**
As a part of debugging https://github.com/NVIDIA/spark-rapids/issues/6431 we found that there is a large performance difference between using a capture group in a regular expression vs a non-capture group. See https://github.com/NVIDIA/spark-rapids/issues/6431#issuecomment-1233236567 for some details on the performance difference.

I don't understand why a capture group would be so expensive if that capture group is never referenced by any sort of a back reference, and the operation being done, like with `matches_re`. 

**Describe the solution you'd like**
I would like to have CUDF automatically detect that a capture group does not need to actually need to capture anything.

**Describe alternatives you've considered**
We do it ourselves manually for any regular expressions that we have hard coded, which is in progress. 
Also when we trans-pile the user's regular expression we could do the same thing I am asking us to do here, but why should only Spark users take advantage of this.

### PR DESCRIPTION
**Pull request by davidwendt**
## Description
Capture groups are used for extracting specific matching substrings but also used for grouping alternation or other sub-pattern matches. If the capture group is not used for extraction then a non-capture group could be specified for these cases. A non-capture group will generate less regex instructions which can help reduce device memory usage.
Since the libcudf strings regex API calls already check where capture groups are required, the API can inform the regex compiler if capture groups are necessary. Then the compiler could automatically convert to non-capture groups reducing the number of instructions produced.
Introduces a new `capture_groups` parameter for use in the regex compiler step for this purpose.
This is an improvement in efficiency and no external behavior has changed.

Also fixes a bug found when testing where a non-capture group pattern is used with an invalid quantifier sequence.
A test case was added to verify the bug fix.

Closes #11663 

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/src/strings/regex/regcomp.cpp [hunk 1]
```diff
@@ -36,7 +36,7 @@ namespace detail {
 namespace {
 // Bitmask of all operators
 #define OPERATOR_MASK 0200
-enum OperatorType {
+enum OperatorType : int32_t {
```
[the remark this answers is not in the document]
Comment (davidwendt) [reply cpp/src/strings/regex/regcomp.cpp:39]: I think this makes sense but I feel this kind of change should go in a separate PR. I don't want to create too many changes here to make it easier to review the intended change. Do you feel this is necessary to approve this change?

**Review by bdice — APPROVED** [review]

File: cpp/src/strings/regex/regcomp.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (bdice) [reply cpp/src/strings/regex/regcomp.cpp:39]: It was nice to fix the newly introduced enum, but the rest can be deferred. Approved now.

File: cpp/src/strings/replace/replace_re.cu [hunk 2]
```diff
@@ -115,7 +115,7 @@ std::unique_ptr<column> replace_re(
   string_view d_repl(replacement.data(), replacement.size());
 
   // compile regex into device object
-  auto d_prog = reprog_device::create(pattern, flags, stream);
+  auto d_prog = reprog_device::create(pattern, flags, capture_groups::NON_CAPTURE, stream);
```
Comment (upsj) [inline cpp/src/strings/replace/replace_re.cu:118]: Just for my understanding: `replace_re` doesn't allow using captured groups in `replacement`, correct?

**Review by upsj — APPROVED** [review]

File: cpp/src/strings/replace/replace_re.cu (hunk 2 above)
Comment (davidwendt) [reply cpp/src/strings/replace/replace_re.cu:118]: That's right. 
There is a [`replace_with_backrefs`](https://docs.rapids.ai/api/libcudf/stable/group__strings__replace.html#ga5e36a14e504147cb4cc8c766fa17f97a) API that does use the captured groups.

**Review by bdice — APPROVED** [review]
Re-approving CMake because GitHub triggered a request after merging in the upstream fix.

**Comment by davidwendt** [comment]
@gpucibot merge

**Pull request merged**